src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,298 @@
+<?php
+
+namespace M6Web\Tornado\Adapter\Swoole;
+
+use Generator;
+use M6Web\Tornado\Adapter\Swoole\Internal\DummyPromise;
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\Promise;
+use Swoole\Coroutine;
+use RuntimeException;
+use Swoole\IDEHelper\StubGenerators\Swoole;
+use Swoole\Runtime;
+use Throwable;
+use function extension_loaded;
+
+class EventLoop implements \M6Web\Tornado\EventLoop
+{
+    private $cids;
+
+    public function __construct()
+    {
+        if (!extension_loaded('swoole')) {
+            throw new RuntimeException(
+                'EventLoop must running only with swoole extension.'
+            );
+        }
+
+        $this->cids = [];
+    }
+
+    private function shiftCoroutine(): mixed
+    {
+        if(count($this->cids) === 0) {
+            return null;
+        }
+
+        $cid = array_shift($this->cids);
+        if(Coroutine::exists($cid)) {
+            Coroutine::resume($cid);
+        } else {
+            $this->shiftCoroutine();
+        }
+
+        return $cid;
+    }
+
+    private function pushCoroutine(): mixed
+    {
+        $cid = Coroutine::getCid();
+        $this->cids[] = $cid;
+        Coroutine::yield();
+        return $cid;
+    }
+
+    private function createPromise(): DummyPromise
+    {
+        return new DummyPromise(function () {
+            $this->shiftCoroutine();
+        });
+    }
+
+    private function resolve($value)
+    {
+        if($value instanceof DummyPromise && !$value->isPending()) {
+            if($value->getException() !== null) {
+                throw $value->getException();
+            }
+
+            $value = $this->resolve($value->getValue());
+        }
+
+        if(is_array($value)) {
+            foreach ($value as $k => $v) {
+                $value[$k] = $this->resolve($v);
+            }
+        }
+
+        return $value;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function wait(Promise $promise): mixed
+    {
+        $promise = DummyPromise::wrap($promise);
+
+        if(count($this->cids) === 0 && $promise->isPending()) {
+            throw new \Error('Impossible to resolve the promise, no more task to execute..');
+        }
+
+        while ($promise->isPending()) {
+            $this->shiftCoroutine();
+        }
```

Comment:
As far as I understand, here your are recoding a scheduler/event-loop… you try to wake-up every coroutine until something is done, but it won't be efficient. I think we should let the Swoole EventLoop do this Job, Tornado just provide an interface over it

Reply:
I did actually end up with this logic with the usage for `DummyPromise` after some tries.
Your proposal logic stay in `Swoole/YieldEventLoop`, but I suspended the work for now.